### PR TITLE
Add BreakOnDistance to DoAfterEventArgs

### LINKED ISSUE #8297: If you start eating food on the floor you can run away to finish eating
## Description
This bug is very beneficial to mice gameplay, it's the current mouse meta

**Screenshots**

https://user-images.githubusercontent.com/31836813/169421711-331606ca-0a8a-4eb3-b744-2d3acd4279ff.mp4

### PR DESCRIPTION
**Pull request by themias**
## About the PR 

There are some interactions where we want to allow movement, but we still want to enforce a distance.
We want people to be able to eat food and move, but want to enforce a maximum distance, to avoid eating food and finishing it while far away.

- BreakOnDistance nullable float to define maximum distance between user and target
- Added optional Used entity in the case of a separate entity being involved between user and target (feeding yourself makes you the user and target. The food is the Used in this case.)
- Coordinates need to be updated in case the target/used has moved. (might be better way of doing this? hopefully not too expensive doing it this way.)
- Need to account for target/used movement for cases like walking while eating food in your hand. Otherwise, BreakOnMove would cancel this.

Fixes #8297

**Screenshots**

1. Eating without moving ✔️ 
2. Eating while moving within range ✔️ 
3. Eating and moving out of range ❌ 
4. Eating held food while walking ✔️ 

<video src=https://user-images.githubusercontent.com/89101928/174168802-ad1c9fde-97c1-459c-8310-73d67ed14834.mp4></video>

**Changelog**

:cl:
- fix: No more Dine and Dash (alt clicking food and running away)

### CODE REVIEW

File: Content.Server/DoAfter/DoAfter.cs [hunk 1]
```diff
@@ -162,6 +164,25 @@ private bool IsCancelled(IEntityManager entityManager)
...
+                    if (!entityManager.GetComponent<TransformComponent>(EventArgs.User).Coordinates.InRange(entityManager, TargetGrid, EventArgs.BreakOnDistance!.Value))
+                        return true;
+                }
+                if (EventArgs.Used != null && !EventArgs.User.Equals(EventArgs.Used))
+                {
+                    //recalculate Used location in case Used has also moved
+                    UsedGrid = entityManager.GetComponent<TransformComponent>(EventArgs.Used!.Value).Coordinates;
+                    if (!entityManager.GetComponent<TransformComponent>(EventArgs.User).Coordinates.InRange(entityManager, UsedGrid, EventArgs.BreakOnDistance!.Value))
+                        return true;
+                }
```
Comment (metalgearsloth) [inline Content.Server/DoAfter/DoAfter.cs:183]: There's a lot of yucky null handling here, also I think the used distance should be irrelevant.

File: Content.Server/DoAfter/DoAfter.cs [hunk 2]
```diff
@@ -20,7 +20,9 @@ public sealed class DoAfter
 
         public EntityCoordinates UserGrid { get; }
 
-        public EntityCoordinates TargetGrid { get; }
+        public EntityCoordinates TargetGrid { get; set; }
+
+        public EntityCoordinates UsedGrid { get; set; }
```
Comment (metalgearsloth) [inline Content.Server/DoAfter/DoAfter.cs:25]: Delete UsedGrid, also why is TargetGrid publicly settable?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
I like the change but food shouldn't have it as this can be used to forcefeed.

**Comment by themias** [comment]
> I like the change but food shouldn't have it as this can be used to forcefeed.

Force feeding still breaks on movement, which is calculated before this break on distance.
In my testing, the user or target moving at all still cancels the feeding.

File: Content.Server/DoAfter/DoAfter.cs (hunk 1 above)
Comment (themias) [reply Content.Server/DoAfter/DoAfter.cs]: We can logically remove a few of these checks, like the target shouldn't be null, and the user shouldn't ever be the used entity. Used is optional, but is needed for this instance where the user and target are the same, and the food on the floor is a separate entity we want to calculate distance from.
We can remove Used entirely if we decide we don't want to use it for cases like that, although it doesn't seem to allow force feeding in this instance.

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**